File: Model/trainModel.py
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
from tensorflow.keras.preprocessing.image import img_to_array, load_img
from cnn import create_model
#import tensorflowjs as tfjs

# Load the CSV file
csv_file_path = 'C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/dataset.csv'
df = pd.read_csv(csv_file_path)

# Load images and labels
def load_data(df, img_size=(128, 128)):
    images = []
    labels = []
    for index, row in df.iterrows():
        img_path = f'C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/images/person/{row["image_name"]}'
        image = load_img(img_path, target_size=img_size)
        image = img_to_array(image) / 255.0  # Normalize the image
        images.append(image)
        labels.append(row['label'])
    
    return np.array(images), np.array(labels)

# Prepare data
X, y = load_data(df)

# Encode labels
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)  # Convert labels to numeric values

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)

# Define input shape for the CNN
input_shape = (128, 128, 3)

# Create and compile the model
model = create_model(input_shape)  # model from cnn.py
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=30, batch_size=32, validation_data=(X_test, y_test))

# Make prediction
y_pred = model.predict(X_test)
y_pred_classes = (y_pred > 0.5).astype(int).flatten()

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred_classes)
print(f"Accuracy: {accuracy}")

# Save the model
model.save('C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/cnn_pose_detection_model.keras')
#tfjs.converters.save_keras_model(model, 'C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/tfjs_model')

File: Model/trainModel.js
import * as tf from '@tensorflow/tfjs-node';
import * as fs from 'fs';
import { createModel } from './cnn.js';

// Load the JSON file
const filePath = 'C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/dataset.json';
const jsonData = JSON.parse(fs.readFileSync(filePath, 'utf8'));

// Load images and labels
async function loadData(jsonData, imgSize = [128, 128]) {
    const images = [];
    const labels = [];
    for (const entry of jsonData) {
        const imgPath = `C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/images/person/${entry.image_name}`;
        const image = await tf.node.decodeImage(fs.readFileSync(imgPath), 3);
        const resizedImage = tf.image.resizeBilinear(image, imgSize);
        const normalizedImage = resizedImage.div(255.0);
        images.push(normalizedImage);
        labels.push(entry.label);
    }
    return [tf.stack(images), labels];
}

// Prepare data
const [X, y] = await loadData(jsonData);

// Encode labels
const labelEncoder = new Map();
const yEncoded = y.map(label => {
    if (!labelEncoder.has(label)) {
        labelEncoder.set(label, labelEncoder.size);
    }
    return labelEncoder.get(label);
});

// Split data
const splitIndex = Math.floor(X.shape[0] * 0.8);
const [X_train, X_test] = tf.split(X, [splitIndex, X.shape[0] - splitIndex]);
const y_train_encoded = yEncoded.slice(0, splitIndex);
const y_test_encoded = yEncoded.slice(splitIndex);
const y_train_tensor = tf.tensor1d(y_train_encoded, 'int32');
const y_test_tensor = tf.tensor1d(y_test_encoded, 'int32');

// Define input shape for the CNN
const inputShape = [128, 128, 3];

// Create and compile the model with Dropout for regularization
const model = createModel(inputShape);
model.compile({
    optimizer: 'adam',
    loss: 'binaryCrossentropy',
    metrics: ['accuracy']
});

// Define the early stopping callback without restoreBestWeights
const earlyStopping = tf.callbacks.earlyStopping({
    monitor: 'val_loss',
    patience: 10, 
    mode: 'min', 
});

// Training the model
await model.fit(X_train, y_train_tensor, {
    epochs: 30,
    batchSize: 32,
    validationData: [X_test, y_test_tensor],
    callbacks: [earlyStopping],
});

// Make prediction
const yPred = model.predict(X_test);
const yPredClasses = yPred.greater(0.5).toInt().dataSync();

// Evaluate the model
const accuracy = tf.metrics.sparseCategoricalAccuracy(y_test_tensor, yPredClasses).dataSync();
console.log(`Accuracy: ${accuracy}`);

// Save the model
await model.save('file://C:/Users/gurjo/Documents/term 8/major project/fitness app/Model/data/cnn_pose_detection_model');